apps/prairielearn/src/ee/lib/lti13.ts

```diff
@@ -499,3 +607,158 @@ export async function fetchRetry(
+
+// https://www.imsglobal.org/spec/lti-nrps/v2p0/#sharing-of-personal-data
+const ContextMembershipSchema = z.object({
+  user_id: z.string(),
+  roles: z.string().array(), // https://www.imsglobal.org/spec/lti/v1p3#role-vocabularies
+  status: z.enum(['Active', 'Inactive']).optional(),
+  email: z.string(), // optional per spec but required for us
+});
+type ContextMembership = z.infer<typeof ContextMembershipSchema>;
+
+export async function updateLti13Scores(
+  assessment_id: string | number,
+  instance: Lti13CombinedInstance,
+  job: ServerJob,
+) {
+  const memberships: ContextMembership[] | null = null;
+  const userSubCache = {};
+
+  // Get the assessment metadata
+  const assessment = await queryRow(
+    sql.select_assessment_with_lti13_course_instance_id,
+    {
+      assessment_id,
+      lti13_instance_id: instance.lti13_instance.id,
+    },
+    AssessmentSchema.extend({
+      lti13_lineitem_id_url: z.string(),
+      lti13_instance_id: z.string(),
+      context_memberships_url: z.string(),
+    }),
+  );
+
+  if (assessment === null) {
+    throw new Error('Invalid assessment.id');
+  }
+
+  job.info(`Sending grade data for ${assessment.tid} ${assessment.title}`);
+
+  const token = await getAccessToken(instance.lti13_instance.id);
+
+  const assessment_instances = await queryRows(
+    sql.select_assessment_instances_for_scores,
+    {
+      assessment_id,
+    },
+    AssessmentInstanceSchema.extend({
+      score_perc: z.number(), // not .nullable() from SQL query
+      date: DateFromISOString, // not .nullable() from SQL query
+      users: UsersWithLti13SubSchema.array(),
+    }),
+  );
+
+  for (const assessment_instance of assessment_instances) {
+    for (const user of assessment_instance.users) {
+      job.info(`ai=${assessment_instance.id}, ${assessment_instance.score_perc}% for ${user.name}`);
+
+      const userId = await lookupSub(user, memberships, instance, userSubCache);
+      if (userId === null) {
+        job.warn(`* Could not find LTI user information for ${user.name} ${user.uid}, skipping...`);
+        continue;
+      }
+
+      // https://canvas.instructure.com/doc/api/score.html#method.lti/ims/scores.create
+      const score: Lti13Score = {
+        timestamp: assessment_instance.modified_at,
+        startedAt: assessment_instance.date,
+        scoreGiven: assessment_instance.score_perc,
+        scoreMaximum: 100,
+        activityProgress: assessment_instance.open ? 'Submitted' : 'Completed',
+        gradingProgress: 'FullyGraded',
+        userId,
+      };
+
+      await fetchRetry(assessment.lti13_lineitem_id_url + '/scores', {
+        method: 'POST',
+        headers: {
+          Authorization: `Bearer ${token}`,
+          'Content-type': 'application/vnd.ims.lis.v1.score+json',
+        },
+        body: JSON.stringify(score),
+      });
+    }
+  }
+}
+
+async function lookupSub(
+  user: UsersWithLti13Sub,
+  memberships: ContextMembership[] | null,
```

Comment:
AFAICT `memberships` isn't updated by the caller, so we'll end up hitting the membership URL repeatedly in a loop. Is that correct? See my other comment about an API that would improve this.